client/lib/parsing.rs

```diff
@@ -281,6 +319,22 @@ pub(super) fn parse_session_info(
     session_version: &str,
     session_entry_point: &str,
 ) -> Result<ExecutableDeployItem> {
+    if !session_args.is_empty() && !session_args_complex.is_empty() {
```

Comment:
I think we need more than just this check for the session args and the `try_conflicting_params` call.

These checks should marry up with the `SessionStrParams::with_` constructors.  For each constructor, we need to check that each of its args which must be non-empty are non-empty, and also any `session_xxx` values which must be  empty are empty.

For example, take `SessionStrParams::with_package_hash`.  We need to ensure `session_package_hash` and `session_entry_point` are non-empty.  `session_args_simple` and `session_args_complex` may both be empty, but both can't be non-empty (as per your check here).  `session_version` is allowed to be empty.  But we also need to ensure that `session_hash`, `session_name`, `session_package_name` and `session_path` are all empty.

Same point for payment ones of course.

Reply:
I've resolved this with the `requires[]` clause in the `check_exactly_one_not_empty!()` macro.

Reply:
This is still not quite right.  Maybe we need a `requires_empty[]` clause for each parameter too?  For example, `(session_path) requires[],` doesn't ensure `session_version` and `session_entry_point` are both empty, which they should be.

I guess this means the tests aren't checking for such conditions either.

Reply:
I added a `requires_empty[]` clause and check, as well as tests for the required variants ensuring that those fail now.